testsuite/tests/typing-jkind-bounds/subsumption/functors.ml

```diff
@@ -0,0 +1,95 @@
+(* TEST
+    flags = "-extension layouts_alpha";
+    expect;
+*)
+
+module F (M : sig type t end) = struct
+  type t : immutable_data with M.t
+end
+
+module Int = struct
+  type t = int
+end
+type t : immutable_data = F(Int).t
+[%%expect {|
+module F :
+  functor (M : sig type t end) -> sig type t : immutable_data with M.t end
+module Int : sig type t = int end
+type t = F(Int).t
+|}]
+
+module T = struct
+  type t
+end
+type t : immutable_data = F(T).t
+[%%expect {|
+module T : sig type t end
+Line 4, characters 0-32:
+4 | type t : immutable_data = F(T).t
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "F(T).t" is immutable_data with T.t
+         because of the definition of t at line 2, characters 2-34.
+       But the kind of type "F(T).t" must be a subkind of immutable_data
+         because of the definition of t at line 4, characters 0-32.
+|}]
+
+module F (M : sig type 'a t end) = struct
+  type 'a t : immutable_data with 'a M.t
+end
+
+module Ref = struct
+  type 'a t = 'a ref
+end
+type 'a t : mutable_data with 'a = 'a F(Ref).t
+[%%expect {|
+module F :
+  functor (M : sig type 'a t end) ->
+    sig type 'a t : immutable_data with 'a M.t end
+module Ref : sig type 'a t = 'a ref end
+type 'a t = 'a F(Ref).t
+|}]
+
+module F (M : sig
+  type t
+  type u
+end) = struct
+  type t : immediate with M.u with M.t
+end
+
+module Int_int = struct
+  type t = int
+  type u = int
+end
+type t : immutable_data = F(Int_int).t
+[%%expect {|
+module F :
+  functor (M : sig type t type u end) ->
+    sig type t : immediate with M.t with M.u end
+module Int_int : sig type t = int type u = int end
+type t = F(Int_int).t
+|}]
+
+module Int_abstract = struct
+  type t = int
+  type u : value mod global
+end
+type t : value mod global = F(Int_abstract).t
+type t : value mod portable with Int_abstract.u = F(Int_abstract).t
+[%%expect {|
+module Int_abstract : sig type t = int type u : value mod global end
+type t = F(Int_abstract).t
+type t = F(Int_abstract).t
+|}]
+
+type t : value mod portable = F(Int_abstract).t
+[%%expect {|
+Line 1, characters 0-47:
+1 | type t : value mod portable = F(Int_abstract).t
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "F(Int_abstract).t" is immediate with Int_abstract.t
```

Comment:
Not for this patch, but this printing is not helpful for users, who might (quite easily) worry about the difference between `immediate` and `value` between the parts of this error message... even though that bit is completely fine.

Reply:
I agree. I have a cr in `sub_jkind_l` that suggests we report which axes disagree, and what the bounds are along those axes.